worker_tasks/categorize.py

```diff
@@ -46,109 +80,349 @@
+        for i, article_1 in enumerate(articles):
+            vertex_attrs["name"].append(article_1)
 
+            for j, article_2 in enumerate(articles):
+                if i == j:
+                    continue
 
-    def is_bigram(self, word):
-        return " " in word
+                if (article_1, article_2) in self._memoized_semantic_relatedness_scores:
+                    relatedness_score = self._memoized_semantic_relatedness_scores[(article_1, article_2)]
+                else:
+                    relatedness_score = self._get_relatedness_score(self._memoized_article_links[article_1], self._memoized_article_links[article_2])
+                    self._memoized_semantic_relatedness_scores[(article_1, article_2)] = relatedness_score
+                    self._memoized_semantic_relatedness_scores[(article_2, article_1)] = relatedness_score
 
+                if relatedness_score:
+                    edges.append([i, j])
+                    edge_attrs["weight"].append(relatedness_score)
 
-    def is_valid_category(self, category):
+        return igraph.Graph(edges=edges, vertex_attrs=vertex_attrs, edge_attrs=edge_attrs)
 
-        # If bigram, check that both words are valid
-        if self.is_bigram(category):
-            category = category.split()
-            return self.is_valid_category(category[0]) and self.is_valid_category(category[1])
 
-        # Ignore single characters
-        if len(category) < 2:
-            return False
+    def _get_relatedness_score(self, article_1_links, article_2_links):
+        """Calculate overlap of two Wikipedia articles: 2 * [# of shared links] / [total # of links]"""
 
-        # Ignore stop words
-        if category.lower() in STOP_WORDS:
+        total_links_count = len(article_1_links) + len(article_2_links)
+
+        if not total_links_count:
+            return 0
+
+        shared_links_count = 0
+        for article_link in article_1_links:
+            weight = 1
+            if "Template" in article_link or \
+                "Wikipedia" in article_link or \
+                "Help" in article_link:
+                weight = 0.1
+            if "Category" in article_link:
+                weight = 1.5
+
+            if article_link in article_2_links:
+                shared_links_count += weight
+
+        return 2.0 * shared_links_count / total_links_count
+
+
+    def _get_best_keywords(self, communities, original_keywords):
+        """Return all keywords from clusters that contain a satisfactory number of the original keywords"""
+
+        best_keywords = set()
+        keyword_counts = Counter(original_keywords)
+
+        for cluster in communities.subgraphs():
+            vertex_names = map(lambda x: x['name'], cluster.vs)
+
+            matching_vertices_count = sum(map(lambda x: keyword_counts[x] if x in keyword_counts else 0, vertex_names))
+
+            cluster_score = float(matching_vertices_count) / len(cluster.vs)
```

Comment:
Won't this almost always be less than 1?

The only way it can't be is if there are words with keyword_counts[word] that are pretty big. Does that happen frequently?